### PR TITLE
Introduce option to detect and fail on path overriding in rule definitions

### PR DESCRIPTION
**Pull request by nicdard**
### Description of changes:

Detect and fail in case of path overriding in rule definition which causes the undefined behavior as described in issue #22. 
Note that in case of OR explicit operators we do not need to fail, as the resulting patterns are being copied and therefore not overridden. 

* Tests are added to `JsonRuleCompiler` as well as`RuleCompiler`.
* The `GenericMachine` class and related builder has been modified to be configured. 
   * By default, the previous behavior is maintained, where the override is happening to avoid breaking compatibility.
* No breaking changes are introduced.

#### Benchmark / Performance (for source code changes):

```
<replace this with output from /src/test/software/amazon/event/ruler/Bechmarks.java here.

The benchmark results can be fetched from "Pull request checks -> Java build -> build (ubuntu-X.Y, 8) -> Run benchmarks".>
```

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by timbray — COMMENTED** [review]
I don't have time to do a thorough review but I don't see anything obviously wrong with the code. The test looks OK, and I hope you've had a careful look at the coverage.  I also didn't see the benchmark output, does this PR have any performance effects?

However, when a rule is supplied via JSON, it makes me very nervous that a JSON object with duplicate keys is accepted. Per RFC8259, such JSON is not interoperable and should not be accepted. Would it be possible to simply reject such JSON rules?

Of course, you could still get conflicting paths by adding rules at the Patterns level, so this is still useful even if you reject JSON with duplicate keys.

**Comment by nicdard** [comment]
@timbray thank you for taking a look! 

> I don't have time to do a thorough review but I don't see anything obviously wrong with the code. The test looks OK, and I hope you've had a careful look at the coverage. 

Test should be covering all the logic, I added enough to test the various cases and corner cases (such the ones described in the issue)

> However, when a rule is supplied via JSON, it makes me very nervous that a JSON object with duplicate keys is accepted. Per RFC8259, such JSON is not interoperable and should not be accepted. Would it be possible to simply reject such JSON rules?

Regarding the above, I think we should still accept it bc of the implicit and operation (I see there is an issue related to the introduction of an new $and operator https://github.com/aws/event-ruler/issues/69), see this example:
https://github.com/aws/event-ruler/issues/22#issuecomment-2827172532

> Of course, you could still get conflicting paths by adding rules at the Patterns level, so this is still useful even if you reject JSON with duplicate keys.

Yes I was thinking that this change just solves all the case where the user could get confused and enter a filter where the semantic is wrong with respect to what was thought to be.


I will run the benchmarks and post the results in this PR

**Comment by timbray** [comment]
> @timbray thank you for taking a look!

First of all, I have to say that I will not approve this PR.  I haven't touched Ruler code in 3 years or more, and I'm not confident enough about not breaking things.   It really needs an insider contributor to look at it.

BTW the former chief maintainer @baldawar has left AWS and moved to Square, so I'm not sure who has ownership of this code on the AWS side.

> Test should be covering all the logic, I added enough to test the various cases and corner cases (such the ones described in the issue)

Not interested in "should be" I would want to hear that (a) you have checked and the coverage % does not go down, and (b) the new code you have written has basically 100% test coverage except for places that obviously don't benefit.

> > However, when a rule is supplied via JSON, it makes me very nervous that a JSON object with duplicate keys is accepted. Per RFC8259, such JSON is not interoperable and should not be accepted. Would it be possible to simply reject such JSON rules?
> 
> Regarding the above, I think we should still accept it bc of the implicit and operation (I see there is an issue related to the introduction of an new $and operator #69), see this example: [#22 (comment)](https://github.com/aws/event-ruler/issues/22#issuecomment-2827172532)

Not sure I understand, but I still don't think Ruler should be accepting broken JSON.

> Regarding the benchmarks, I saw this line in the [PR template](https://github.com/aws/event-ruler/blob/main/.github/PULL_REQUEST_TEMPLATE.md#benchmark--performance-for-source-code-changes):
> 
> > The benchmark results can be fetched from "Pull request checks -> Java build -> build (ubuntu-X.Y, 8) -> Run benchmarks".>
> 
> Could you please approve the workflow to run so that I can get the results. If I run the benchmarks in my machine, I'll get a different result (double or more the speed).

I approved it but it failed because the git actions are out of date, some Ubuntu version thing.  If you want to do another PR to fix that, it is the kind of thing I would approve.

> as expected the performance is quite similar since the change is only affecting the compilation of the filter. Should I post the change where I substitute the direct constructor invocation of `Machine` class with the builder call?

**Comment by nicdard** [comment]
> First of all, I have to say that I will not approve this PR. I haven't touched Ruler code in 3 years or more, and I'm not confident enough about not breaking things. It really needs an insider contributor to look at it.

Thank you for, makes sense no worries.

> BTW the former chief maintainer @baldawar has left AWS and moved to Square, so I'm not sure who has ownership of this code on the AWS side.

I see thanks for the info.

> > Test should be covering all the logic, I added enough to test the various cases and corner cases (such the ones described in the issue)
> 
> Not interested in "should be" I would want to hear that (a) you have checked and the coverage % does not go down, and (b) the new code you have written has basically 100% test coverage except for places that obviously don't benefit.

I measured the coverage, and there is no change in it. all the new branches are covered in the code.

> > > However, when a rule is supplied via JSON, it makes me very nervous that a JSON object with duplicate keys is accepted. Per RFC8259, such JSON is not interoperable and should not be accepted. Would it be possible to simply reject such JSON rules?
> > 
> > 
> > Regarding the above, I think we should still accept it bc of the implicit and operation (I see there is an issue related to the introduction of an new $and operator #69), see this example: [#22 (comment)](https://github.com/aws/event-ruler/issues/22#issuecomment-2827172532)
> 
> Not sure I understand, but I still don't think Ruler should be accepting broken JSON.

I see your point, maybe it can be addressed separately given the above discussion I was mentioning.

> > Regarding the benchmarks, I saw this line in the [PR template](https://github.com/aws/event-ruler/blob/main/.github/PULL_REQUEST_TEMPLATE.md#benchmark--performance-for-source-code-changes):
> > > The benchmark results can be fetched from "Pull request checks -> Java build -> build (ubuntu-X.Y, 8) -> Run benchmarks".>
> > 
> > 
> > Could you please approve the workflow to run so that I can get the results. If I run the benchmarks in my machine, I'll get a different result (double or more the speed).
> 
> I approved it but it failed because the git actions are out of date, some Ubuntu version thing. If you want to do another PR to fix that, it is the kind of thing I would approve.

Thank you! I can see if I have time I can open a new PR :)


> > as expected the performance is quite similar since the change is only affecting the compilation of the filter. Should I post the change where I substitute the direct constructor invocation of `Machine` class with the builder

**Comment by timbray** [comment]
Hey @embano1 is there anyone at AWS who should be looking at this?

**Comment by embano1** [comment]
Pinged some folks internally.

**Comment by rishibaldawa** [comment]
hey @embano1, 👋  I'd recommend starting w @kamouney @scabraha . 

@nicdard will look at this PR later today.

File: src/main/software/amazon/event/ruler/GenericMachine.java [hunk 1]
```diff
@@ -764,19 +764,51 @@ public static class Builder<M extends GenericMachine<T>, T> {
...
+        private boolean ruleOverriding = true;
+
         Builder() {}
 
         public Builder<M,T> withAdditionalNameStateReuse(boolean additionalNameStateReuse) {
             this.additionalNameStateReuse = additionalNameStateReuse;
             return this;
         }
 
+        public Builder<M, T> withRuleOverriding(boolean ruleOverriding) {
```
Comment (rishibaldawa) [inline src/main/software/amazon/event/ruler/GenericMachine.java:801]: withOverridesForDuplicateRules might be more clearer.

File: src/main/software/amazon/event/ruler/JsonRuleCompiler.java [hunk 2]
```diff
@@ -343,7 +380,12 @@ private static void writeRules(final List<Map<String, List<Patterns>>> rules,
         if (rules.isEmpty()) {
             rules.add(new HashMap<>());
         }
-        rules.forEach(rule -> rule.put(name, values));
+        rules.forEach(rule -> {
+            if (!withOverriding && rule.containsKey(name)) {
+                throw new IllegalArgumentException("Overriding path '" + name + "' is already defined. Please check that the rule does not list the same path multiple times.");
```
Comment (rishibaldawa) [inline src/main/software/amazon/event/ruler/JsonRuleCompiler.java:385]: I think we use barf() for error handling. Also, wouldn't recommend putting in the name in the exception message in case there's something sensitive that folks don't want to be logged.

File: src/main/software/amazon/event/ruler/RuleCompiler.java [hunk 3]
```diff
@@ -240,6 +274,9 @@ private static void writeRules(final Map<String, List<Patterns>> rule,
         if (values.isEmpty()) {
             barf(parser, "Empty arrays are not allowed");
         }
+        if (!withOverriding && rule.containsKey(name)) {
+            throw new IllegalArgumentException("Overriding path '" + name + "' is already defined. Please check that the rule does not list the same path multiple times.");
```
Comment (rishibaldawa) [inline src/main/software/amazon/event/ruler/RuleCompiler.java:278]: Same comments as in the other file.

File: src/main/software/amazon/event/ruler/JsonRuleCompiler.java (hunk 2 above)
Comment (rishibaldawa) [reply src/main/software/amazon/event/ruler/JsonRuleCompiler.java:385]: Error msg can be `barf("Same path cannot be allowed multiple times")` for consistency w rest of the messages

**Comment by ghost** [comment]
also @nicdard , TY 

Screenshot 2025-05-14 at 5 10 56 PM

File: src/main/software/amazon/event/ruler/JsonRuleCompiler.java (hunk 2 above)
Comment (nicdard) [reply src/main/software/amazon/event/ruler/JsonRuleCompiler.java]: oh thanks good point, I'll use barf as well.
however in terms of sensitive info, I guess barf is worst since it's going to print all the source as a string:

From `JsonProcessingException`:
```java
@Override public String getMessage() {
        String msg = super.getMessage();
        if (msg == null) {
            msg = "N/A";
        }
        JsonLocation loc = getLocation();
        String suffix = getMessageSuffix();
        // mild optimization, if nothing extra is needed:
        if (loc != null || suffix != null) {
            StringBuilder sb = new StringBuilder(100);
            sb.append(msg);
            if (suffix != null) {
                sb.append(suffix);
            }
            if (loc != null) {
                sb.append('\n');
                sb.append(" at ");
                sb.append(loc.toString());
            }
            msg = sb.toString();
        }
        return msg;
    }
```

If it's ok I would state the path in the message, just bc it's more clear which one is the offending path. wdyt?

File: src/main/software/amazon/event/ruler/GenericMachine.java (hunk 1 above)
Comment (nicdard) [reply src/main/software/amazon/event/ruler/GenericMachine.java]: refactored :)

**Comment by nicdard** [comment]
@rishibaldawa thank you for your review, all the changes are in this commit: https://github.com/aws/event-ruler/pull/212/commits/46aced22a208da2211fbf35d60bf83826c04d526

I will re-order the commits if needed to have the change of version as last. don't know if you do squash and merge than we don't need that. leaving them in this order so you can more easily follow up in the review for now :)

File: src/main/software/amazon/event/ruler/JsonRuleCompiler.java (hunk 2 above)
Comment (ghost) [reply src/main/software/amazon/event/ruler/JsonRuleCompiler.java]: sounds good to me though I'll let AWS make the final call if this is acceptable from a security standpoint.

**Comment by ghost** [comment]
We used to squash and merge. I'd imagine we'd do the same going forward unless there's a good reason.

**Comment by shikhartanwar** [comment]
@rishibaldawa Who would be the right person to reach out to for getting this PR merged?

**Comment by ghost** [comment]
@kamouney should be the one.

**Comment by shikhartanwar** [comment]
> @kamouney should be the one.

@kamouney Could you help us with the approval and merging the PR?

**Comment by shikhartanwar** [comment]
> @kamouney Could you help us with the approval and merging the PR?

Thanks a lot for the approval @kamouney. I see that all checks have passed now. Could you please merge it and generate a release?

**Comment by shikhartanwar** [comment]
> Thanks a lot for the approval @kamouney. I see that all checks have passed now. Could you please merge it and generate a release?

@kamouney Sorry for bothering you multiple times, but can you please generate a latest release/tag that has this fix.

**Comment by nicdard** [comment]
> > Thanks a lot for the approval @kamouney. I see that all checks have passed now. Could you please merge it and generate a release?
> 
> @kamouney Sorry for bothering you multiple times, but can you please generate a latest release/tag that has this fix.

+1  this would allow us to use the new version :)

**Comment by shikhartanwar** [comment]
> > > Thanks a lot for the approval @kamouney. I see that all checks have passed now. Could you please merge it and generate a release?
> > 
> > 
> > @kamouney Sorry for bothering you multiple times, but can you please generate a latest release/tag that has this fix.
> 
> +1 this would allow us to use the new version :)

@kamouney Any update on this please?

**Comment by kamouney** [comment]
> > > > Thanks a lot for the approval @kamouney. I see that all checks have passed now. Could you please merge it and generate a release?
> > > 
> > > 
> > > @kamouney Sorry for bothering you multiple times, but can you please generate a latest release/tag that has this fix.
> > 
> > 
> > +1 this would allow us to use the new version :)
> 
> @kamouney Any update on this please?

Sorry for the delay, I released [V1.8.2 ](https://github.com/aws/event-ruler/releases/tag/v1.8.2)

**Comment by nicdard** [comment]
> > > > > Thanks a lot for the approval @kamouney. I see that all checks have passed now. Could you please merge it and generate a release?
> > > > 
> > > > 
> > > > @kamouney Sorry for bothering you multiple times, but can you please generate a latest release/tag that has this fix.
> > > 
> > > 
> > > +1 this would allow us to use the new version :)
> > 
> > 
> > @kamouney Any update on this please?
> 
> Sorry for the delay, I released [V1.8.2 ](https://github.com/aws/event-ruler/releases/tag/v1.8.2)

Thank you!
